lib/application/widgets/theme/rich_text_field.dart

```diff
@@ -0,0 +1,436 @@
+import 'dart:convert';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+import 'package:flutter_quill/flutter_quill.dart' as quill;
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:layoutr/common_layout.dart';
+import 'package:memo/application/constants/dimensions.dart' as dimens;
+import 'package:memo/application/constants/images.dart' as images;
+import 'package:memo/application/constants/strings.dart' as strings;
+import 'package:memo/application/theme/theme_controller.dart';
+import 'package:memo/application/utils/bottom_sheet.dart';
+import 'package:memo/application/widgets/material/asset_icon_button.dart';
+import 'package:memo/application/widgets/theme/custom_button.dart';
+import 'package:memo/application/widgets/theme/themed_container.dart';
+import 'package:tuple/tuple.dart';
+
+@immutable
+class RichTextEditingValue {
+  const RichTextEditingValue({
+    this.richText = '',
+    this.plainText = '',
+    this.selection = const TextSelection.collapsed(offset: -1),
+  });
+
+  final String richText;
+  final String plainText;
+  final TextSelection selection;
+
+  RichTextEditingValue copyWith({String? richText, String? plainText, TextSelection? selection}) =>
+      RichTextEditingValue(
+        richText: richText ?? this.richText,
+        plainText: plainText ?? this.plainText,
+        selection: selection ?? this.selection,
+      );
+}
+
+class RichTextFieldController extends ValueNotifier<RichTextEditingValue> {
+  RichTextFieldController({
+    String? richText,
+    String? plainText,
+    TextSelection? selection,
+  }) : super(
+          RichTextEditingValue(
+            richText: richText ?? '',
+            plainText: plainText ?? '',
+            selection: selection ?? const TextSelection.collapsed(offset: -1),
+          ),
+        );
+
+  String get richText => value.richText;
+  String get plainText => value.plainText;
+  TextSelection get selection => value.selection;
+
+  set richText(String newValue) => value = value.copyWith(richText: newValue);
+  set plainText(String newValue) => value = value.copyWith(plainText: newValue);
+  set selection(TextSelection newValue) => value = value.copyWith(selection: newValue);
+}
```

Comment:
Again, I may be out of the loop here, but what does all this "RichTextFieldController" does that a common "TextFieldController" doesn't? Can't we modify the original one instead of creating a whole new one, with custom hooks and such?

Reply:
Actually I started using a common `TextEditingController`. The problems that I found during its usage are:

1. We would have to use `text` to store the string version of the rich content, which is a little bit counter intuitive, since the `text` normally stores plain text when using `TextEditingController`.
2. The `TextEditingController` only supports storing two properties: `text` (String) and `selection` (TextSelection), but we also need to have access to the plain text version of the rich text editor. Because of that necessity and further ones that may happen, I think that having a custom controller gives us flexibility to add/remove necessary information.